### ISSUE TITLE
brms::log_lik creating zombies if cmdstanr is loaded

### ISSUE THREAD

**Issue by avehtari** [paul-buerkner/brms#1658]
Originally issue reported at https://github.com/n-kall/priorsense/issues/48, but now reporting here as this has been narrowed done to `brms`

Minimal example by @n-kall
```
m <- brms::brm(yield ~ 1, npk, backend = "rstan")
brms::log_lik(m) # no zombies

library(cmdstanr)
brms::log_lik(m) # creates zombies
```

I checked that the zombies are created by the line https://github.com/paul-buerkner/brms/blob/298b947fa9cfb914aeb7cb3aab7974aa179682b1/R/log_lik.R#L127

**Comment by paul-buerkner** [comment]
Thank you for reporting this issue! Do you or anyone else have any clue why this is happening?

**Comment by avehtari** [comment]
Only idea is that loading `cmdstanr` is changing the behavior of `plapply()` or changing how `log_lik_fun` or `object` are created

**Comment by avehtari** [comment]
@andrjohns found the problem in definition of `plapply()` and will make a PR

**Comment by paul-buerkner** [comment]
Great! Thank you!

**Comment by paul-buerkner** [comment]
In case @andrjohns is busy. If you can tell me what the issue is, I am also happy to fix it myself.

**Comment by avehtari** [comment]
@andrjohns has a branch https://github.com/andrjohns/brms/tree/brms-zombies
and I and Noa didn't see zombies with that branch

**Comment by paul-buerkner** [comment]
Thanks! I will just copy over the code from him and credit you and him for the fix. Okay?

**Comment by paul-buerkner** [comment]
I have fixed it accordingly. I also do not see any Zombies in the minimal example anymore. Thank you all for fixing this!

**Comment by n-kall** [comment]
Last time I tested (Friday) I was still getting zombies from @andrjohns branch. It appears to occur on Ubuntu but no macOS

**Comment by paul-buerkner** [comment]
Interesting. I was only testing on windows and macOS. Let's open it again and hope that we can fix this too. Any more ideas what's going on?

**Comment by n-kall** [comment]
I just confirmed that I still get zombies with the current master brms version (using the minimal example) on Ubuntu 20.04. No idea why though, but somehow related to cmdstanr being loaded

**Comment by avehtari** [comment]
This issue has been closed, but the fix was reverted and with the latest `brms` from github, I still get zombies

**Comment by paul-buerkner** [comment]
Good point. Will reopen this issue.

**Comment by emstruong** [comment]
I don't know how to make a simple reprex of this or if this should be considered a separate issue or if this is even a `brms issue`, but on `Ubuntu 22.04` I've been using `brms` in a Monte Carlo Simulation via `library(SimDesign)` recently. 

Something I've noticed is that as time goes by, the memory usage keeps on rocketing up and there seems to be a persistent error message that `parallel:::.workRSOCK()` is not able to connect to the localhost's port. This is despite that port being explicitly unblocked to `localhost` AFAICT via `ufw`. (I'm aware that parallel::makeCluster() uses ports 11000:11999). There sometimes appears to be many zombie `.workRSOCK()` R processes floating around in `htop`, each taking up a lot of RAM. 

I don't know if using the `future` backend would work or if I should just use `rstanarm` for my simulation study. I figured I should just leave this note here in-case anyone else is doing something similar because I failed to find anything similar/helpful through Google.

**Comment by paul-buerkner** [comment]
Does the same problem happen when you use the rstan (vs. cmdstanr) backend?

**Comment by emstruong** [comment]
> Does the same problem happen when you use the rstan (vs. cmdstanr) backend?

I only used `rstan`, so I guess that is a bit different from the above.

Here's the R syntax that I'm running on each independent dataset

```R
brms::brm(
        formula = grade ~ 1 + (1 | classroom),
        data = dat,
        iter = setiter,
        chains = setchains,
        warmup = floor(setiter / 2),
        control = list(adapt_delta = 0.8, max_treedepth = 10),
        # Default Arguments, manually set in case brms api changes
        cores = setcores,
        prior = NULL,
        silent = 2,
        refresh = 0,
        open_progress = FALSE
      )
```

**Comment by avehtari** [comment]
Any progress with this? We're going to ask students to use priorsense, and I'm worried about the zombies (clearing them away from my own laptop every now and then is a nuisance)

**Comment by paul-buerkner** [comment]
not from my side. I just have no idea why this is happening...

Aki Vehtari ***@***.***> schrieb am Di., 12. Nov. 2024, 21:13:

**Comment by paul-buerkner** [comment]
perhaps this weekend I can look at this again

Paul Buerkner ***@***.***> schrieb am Di., 12. Nov. 2024,
22:01:

**Comment by n-kall** [comment]
> We're going to ask students to use priorsense, and I'm worried about the zombies

Doesn't this also occur when using loo? log_lik is called there also

**Comment by avehtari** [comment]
As the above minimal example doesn't currently produce zombies for me, here is a new minimal example
```
library(brms)
options(brms.backend = "cmdstanr", mc.cores = 2)

m0 <- brms::brm(yield ~ 0, npk, refresh=0)
ll<-brms::log_lik(m0) # no zombies

m1 <- brms::brm(yield ~ 1, npk, refresh=0)
ll<-brms::log_lik(m1) # yes zombies
```

**Comment by emstruong** [comment]
For what it's worth, on `brms version 2.22.6` and `Ubuntu 22.04`, I still get the `Warning message: In socketConnection(master, port = port, blocking = TRUE, open = "a+b",  :   localhost:11049 cannot be opened`. (The particular port number shouldn't matter.) In the past, after a while, I would observe multiple R zombie processes spawn over time following these Warning messages. Whether or not there's a causal relationship is unknown to me.

However, the warning message doesn't happen with every invocation of `brm`, so I'm going to guess that the bug is more complex. I currently speculate that this bug is due to some unfortunate interaction between how `SimDesign` handles parallel processing and how `brm` handles parallel processing. 

I'll open a separate issue if I can get a reliable reprex.

**Comment by paul-buerkner** [comment]
Thank you!
